groups/ais/instance.tf

```diff
@@ -21,6 +21,14 @@ resource "aws_security_group" "common" {
     cidr_blocks = var.deployment_cidrs
   }
 
+  ingress {
+    description = "Allow Informix HDR connectivity for Visual Basic services"
+    from_port   = 6278
+    to_port     = 6278
+    protocol    = "TCP"
+    cidr_blocks = local.visual_basic_app_cidrs
+  }
```

Comment:
Ideally this would be added as a new `aws_security_group_rule` resource that loops through the concatenated list.

Reply:
I believe the recommended approach now is to use [aws_vpc_security_group_egress_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_egress_rule) and [aws_vpc_security_group_ingress_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_security_group_ingress_rule) resources for all new rules, with the caveat that those resources should not be used in combination with any `aws_security_group` resource with inline `ingress`/`egress` rules, or with associated `aws_security_group_rule` resources. It had cross my mind here, and reworking it for this single security group shouldn't be too much effort.

It may require more effort to adopt that approach with other Tuxedo stacks, particularly where we're using a combination of `for_each`, multiple `dynamic` blocks, and large rule sets, e.g. [aws_security_group.services](https://github.com/companieshouse/chl-tuxedo-stack/blob/infrastructure/groups/frontend/instance.tf#L12-L95) in  [chl-tuxedo-stack](https://github.com/companieshouse/chl-tuxedo-stack/blob/infrastructure/groups/frontend/instance.tf#L12-L95). That will obviously need to be addressed separately, but it would be good for consistency sake to use the same approach there as we do here.

Reply:
While I appreciate `aws_vpc_security_group_egress_rule` and `aws_vpc_security_group_ingress_rule` are now the recommended approaches, this isn't the approach we've reviewed and agreed upon. We should endeavour to update our practices accordingly, of course, as things evolve.

Reply:
I think it would be a good time to have a discussion around whether our best practice documentation should be treated as a _living document_ or should be subject to some form of versioning or periodic updates, given this has only recently been introduced.

This is admittedly a relatively small change here, but it does highlight the risk we may introduce of adopting practices that are already behind the upstream recommendations, knowing that we will have to rework things again in future. I'll mull it over some more, but I may opt for the `aws_security_group_rule` approach here to keep things simple. The example above might be a better illustration where this could hurt us, productivity-wise at least: [aws_security_group.servides](https://github.com/companieshouse/chl-tuxedo-stack/blob/infrastructure/groups/frontend/instance.tf#L12-L95). Ideally we want to avoid having to rework that more than is necessary, given the scope and complexity involved.